parlai/crowdsourcing/tasks/turn_annotations/frontend/components/chat_app_with_onboarding.jsx

```diff
@@ -0,0 +1,230 @@
+    },
+  });
+
+  let {
+    blockedReason,
+    blockedExplanation,
+    taskConfig,
+    isPreview,
+    previewHtml,
+    isLoading,
+    agentId,
+    handleSubmit,
+    connect,
+    destroy,
+    sendMessage,
+    isOnboarding,
+    agentState,
+    agentStatus,
+  } = mephistoProps;
+
+  React.useEffect(() => {
+    if (agentId) {
+      console.log("connecting...");
+      connect(agentId);
+    }
+  }, [agentId]);
+
+  React.useEffect(() => {
+    if (isOnboarding && agentStatus === AGENT_STATUS.WAITING) {
+      handleSubmit();
+    }
+  }, [isOnboarding, agentStatus]);
+
+  React.useEffect(() => {
+    // clear messages when onboarding changes status
+    addMessage(false);
+  }, [isOnboarding])
+
+  const handleMessageSend = React.useCallback(
+    (message) => {
+      message = {
+        ...message,
+        id: agentId,
+        episode_done: agentState?.task_done || false,
+      };
+      return sendMessage(message)
+        .then(addMessage)
+        .then(() => setInputMode(INPUT_MODE.WAITING));
+    },
+    [agentId, agentState?.task_done, addMessage, setInputMode]
+  );
+
+  if (blockedReason !== null) {
+    return <h1>{blockedExplanation}</h1>;
+  }
+  if (isLoading) {
+    return <div>Initializing...</div>;
+  }
+  if (isPreview) {
+    if (!taskConfig.has_preview) {
+      return <TaskPreviewView description={taskConfig.task_description} />;
+    }
+    if (previewHtml === null) {
+      return <div>Loading...</div>;
+    }
+    return <div dangerouslySetInnerHTML={{ __html: previewHtml }} />;
+  }
+  if (isOnboarding) {
+    return <OnboardingComponent 
+      onboardingData={taskConfig.onboarding_data} 
+      annotationBuckets={taskConfig.annotation_buckets} 
+      annotationQuestion={taskConfig.annotation_question} 
+      onSubmit={(dat) => {
+        handleMessageSend({text: '', task_data: dat});
+      }} 
+    />;
+  }
+
+  return (
+    <MephistoContext.Provider value={mephistoProps}>
+      <AppContext.Provider
+        value={{
+          taskContext,
+          appSettings,
+          setAppSettings,
+          onTaskComplete: () => {
+            destroy();
+            handleSubmit({});
+          },
+        }}
+      >
+        <div className="container-fluid" id="ui-container" style={{'display': "block"}}>
+          <BaseFrontend
+            inputMode={inputMode}
+            messages={messages}
+            onMessageSend={handleMessageSend}
+            renderMessage={renderMessage}
+            renderSidePane={renderSidePane}
+            renderTextResponse={renderTextResponse}
+            renderResponse={renderResponse}
+          />
+        </div>
+      </AppContext.Provider>
+    </MephistoContext.Provider>
+  );
+}
+
+function TaskPreviewView({ description }) {
+  return (
+    <div className="preview-screen">
+      <div
+        dangerouslySetInnerHTML={{
+          __html: description,
+        }}
+      />
+    </div>
+  );
+}
+
+export { CustomOnboardingChatApp, AppContext };
```

Comment:
Fine for now, but this entire file seems like boilerplate that ought not to have to be copy and pasted into a given task no?

Reply:
Agreed, I think `CustomOnboardingChatApp` or simillar may want to be a part of `bootstrap-chat` - @pringshia we should discuss a few common patterns for abstracting this.